# Has flying in Europe become too cheap to care about?

---

## Introduction

In 1997, the European Union finished opening up its aviation market. From that point, any EU airline could fly any route between member states at any price. Within a decade, a handful of carriers, Ryanair, easyJet, and later Wizz Air, had built an entirely new kind of airline around this freedom. They flew between smaller secondary airports, charged for almost every extra service, and sold tickets at prices that older, established carriers simply could not match. By the late 2000s, a return flight from Amsterdam to Barcelona cost less than a train ticket across the Netherlands.

These low-cost carriers, often called LCCs, grew quickly. By 2019 they held more than half of all seats on flights within Europe, a share that was close to zero in 1995. Ryanair alone carried over 150 million passengers that year, making it the largest international airline in the world by passenger numbers. Together with easyJet and Wizz Air, LCCs fundamentally changed who could afford to fly, and where.

The effects of this change are genuinely contested. On one side, there is a strong case that LCCs made flying accessible to people across Europe who previously could not afford it, opening up routes, especially to Central and Eastern Europe, that simply did not exist at reasonable prices before. On the other side, there is an equally strong case that this boom in cheap flying came with a serious environmental price, driving up CO2 emissions from European aviation year after year.

This data story examines both sides of that question using official European aviation statistics and emissions data measured by EUROCONTROL, the organisation that manages European airspace.

### Perspective 1 - LCCs made flying accessible to far more people

Before LCCs, flying within Europe was expensive. National airlines such as Air France, Lufthansa, and Alitalia kept fares high enough that for many people, flying was an occasional luxury rather than an everyday option. LCCs changed this in two ways. First, they dramatically cut average fares on routes where they competed, forcing established carriers to lower their own prices. Second, they opened entirely new routes, particularly between Western Europe and the newly enlarged EU member states in Central and Eastern Europe, that had never been served by affordable flights before.

- Argument 1.1: Total EU passenger numbers grew by roughly 46% between 2008 and 2019, far faster than population or income growth alone could explain. This is not a fixed group of travellers moving between airlines. The overall market grew substantially.
- Argument 1.2: The growth was concentrated in exactly the places where LCCs were most active: Romania, Hungary, Poland, Ireland, and the Baltic states. These are countries where cheap scheduled flights simply did not exist for most people before Ryanair and Wizz Air arrived.
- Argument 1.3: Airports that LCCs used as their bases grew at about double the rate of major traditional hubs between 2004 and 2019. Importantly, the traditional hubs also kept growing, which means LCCs were not simply taking passengers away from existing routes. They were bringing in new ones.

### Perspective 2 - The boom in cheap flying has a real environmental cost

Flying produces CO2. More flights mean more CO2, regardless of whether the ticket cost twenty euros or two hundred. The short routes that LCCs mostly operate are also the least fuel-efficient kind of flying: a large share of total fuel burn happens during take-off and climb, so shorter flights have higher emissions per kilometre than longer ones. LCCs fill their aircraft well, which helps, but the sheer number of extra flights they generate means the total environmental cost keeps rising.

- Argument 2.1: European aviation CO2 emissions rose from 163 million tonnes in 2010 to 211 million tonnes in 2019, a rise of 29%. This is not an estimate. It is measured from actual flight data by EUROCONTROL.
- Argument 2.2: The amount of CO2 produced per flight did not go down over this period. It actually went up slightly, meaning improvements in aircraft technology were not enough to offset the growth in flights.
- Argument 2.3: The growth in CO2 tracked the growth in passengers almost exactly, year by year. Every wave of new passengers that LCCs brought into the market corresponded to a near-identical wave of new emissions.

### A note on the data

Four datasets are used in this story. Three come from Eurostat (the EU statistics office) and cover quarterly passenger numbers and flight movements by country, and annual passenger figures for individual airports. The fourth is the EUROCONTROL CO2 emissions dataset, which gives actual measured emissions and flight movements for 44 European states, month by month, from 2010 to 2025.

None of the datasets separately label flights as LCC or traditional carrier flights. To compare the two groups, airports are classified as either LCC-focused or traditional-carrier-focused based on which airline dominates their traffic. This is a simplification, and its limits are noted wherever it is used.


---

## Dataset and Preprocessing

### Datasets

| Dataset | Source | What it contains | Time range |
|---|---|---|---|
| avia_paoc | Eurostat | Quarterly passengers boarded, by country | 1993-2025, 39 countries |
| avia_tf_acc | Eurostat | Quarterly passenger flight movements, by country | 2007-2026, 39 countries |
| avia_paoa | Eurostat | Annual passengers, by airport | 2004-2025, ~350 airports |
| co2_emmissions_by_state_YYYY.csv | EUROCONTROL | Monthly CO2 (tonnes) and flight movements, by state | 2010-2025, 44 states |

The EUROCONTROL CO2 figures come from actual flight records using ICAO calculation methods. Four states (Denmark, Norway, Portugal, United Kingdom) are marked as partially estimated in the source data. They are kept in the analysis because removing four of Europe's biggest aviation countries would distort the European total more than leaving in the estimates.

The year 2026 is excluded from the CO2 analysis as only January through April data is available.

### Airport classification

Because the Eurostat data does not separate LCC from traditional carrier traffic, airports are classified by which type of carrier dominates their traffic. The table below shows the airports used.

| Airport | Code | Category | Main carrier |
|---|---|---|---|
| London Stansted | UK_EGSS | LCC-focused | Ryanair |
| London Luton | UK_EGGW | LCC-focused | easyJet |
| Liverpool | UK_EGGP | LCC-focused | Ryanair |
| East Midlands | UK_EGNX | LCC-focused | Ryanair / Wizz Air |
| Bristol | UK_EGGD | LCC-focused | easyJet |
| Brussels South Charleroi | BE_EBCI | LCC-focused | Ryanair |
| Frankfurt Hahn | DE_EDFH | LCC-focused | Ryanair |
| Girona | ES_LEGE | LCC-focused | Ryanair |
| Bergamo Orio al Serio | IT_LIME | LCC-focused | Ryanair |
| Bologna | IT_LIPE | LCC-focused | Ryanair / Wizz Air |
| Dublin | IE_EIDW | LCC-focused | Ryanair |
| Lyon | FR_LFLL | LCC-focused | easyJet |
| Rotterdam | NL_EHRD | LCC-focused | Transavia |
| London Heathrow | UK_EGLL | Traditional | British Airways |
| London Gatwick | UK_EGKK | Traditional | British Airways |
| Paris CDG | FR_LFPG | Traditional | Air France |
| Frankfurt | DE_EDDF | Traditional | Lufthansa |
| Amsterdam Schiphol | NL_EHAM | Traditional | KLM |
| Madrid Barajas | ES_LEMD | Traditional | Iberia |
| Rome Fiumicino | IT_LIRF | Traditional | Alitalia / ITA |
| Munich | DE_EDDM | Traditional | Lufthansa |
| Zurich | CH_LSZH | Traditional | Swiss |
| Vienna | AT_LOWW | Traditional | Austrian |

Note that Gatwick also has a large easyJet operation, and Dublin handles Aer Lingus alongside Ryanair. These airports are kept in their main category because one type still dominates, but readers should know the boundary is not perfectly clean.

### Preprocessing steps

1. For all three Eurostat files: keep only the relevant measurement columns, remove rows with missing values, and parse the quarterly time codes into year and quarter numbers. Sum the four quarters to get annual totals.
2. For the CO2 files: load all 16 annual files (2010-2025) and combine them. The files from 2010-2018 are missing the FLIGHT_MONTH column that appears in later files. That column is dropped during the merge since only year, month, state name, CO2 quantity, and flight movements are needed. Totals are summed by year.
3. For the indexed growth charts: both series are divided by their value in the base year (2013, the first year where both the CO2 data and EU27 passenger data overlap) and multiplied by 100.
4. For the choropleth map: Eurostat uses non-standard country codes (EL for Greece, UK for the United Kingdom). These are converted to the three-letter ISO codes that Plotly needs.


In [1]:
# Cell 1: Imports
import pandas as pd
import numpy as np
import plotly.graph_objects as go
import plotly.express as px
from plotly.subplots import make_subplots
import os
import warnings
warnings.filterwarnings('ignore')

# Colour palette - IBM colour-blind safe
COL_LCC    = '#648FFF'  # Blue
COL_TRAD   = '#FE6100'  # Orange
COL_CO2    = '#DC267F'  # Pink
COL_EFF    = '#785EF0'  # Purple
COL_GREY   = '#AAAAAA'
COL_GRID   = '#eeeeee'

print("Imports OK.")


Imports OK.


In [2]:
# Cell 2: Load and clean the three Eurostat datasets

# avia_paoc: quarterly passengers boarded, by country
raw_paoc = pd.read_csv('avia_paoc__custom_21854155_quarterly_linear_2_0.csv')
df_paoc = (
    raw_paoc[raw_paoc['tra_meas'] == 'PAS_BRD']
    [['geo', 'Geopolitical entity (reporting)', 'TIME_PERIOD', 'OBS_VALUE']]
    .rename(columns={
        'geo': 'country_code',
        'Geopolitical entity (reporting)': 'country_name',
        'OBS_VALUE': 'passengers'
    })
    .dropna(subset=['passengers'])
    .copy()
)
df_paoc['year']    = df_paoc['TIME_PERIOD'].str[:4].astype(int)
df_paoc['quarter'] = df_paoc['TIME_PERIOD'].str[-1].astype(int)
df_paoc_annual = df_paoc.groupby(
    ['country_code', 'country_name', 'year'], as_index=False
)['passengers'].sum()

# avia_tf_acc: quarterly passenger flight movements, by country
raw_tf = pd.read_csv('avia_tf_acc__custom_21854438_quarterly_linear_2_0.csv')
df_tf = (
    raw_tf[raw_tf['tra_meas'] == 'CAF_PAS']
    [['geo', 'Geopolitical entity (reporting)', 'TIME_PERIOD', 'OBS_VALUE']]
    .rename(columns={
        'geo': 'country_code',
        'Geopolitical entity (reporting)': 'country_name',
        'OBS_VALUE': 'flights'
    })
    .dropna(subset=['flights'])
    .copy()
)
df_tf['year']    = df_tf['TIME_PERIOD'].str[:4].astype(int)
df_tf['quarter'] = df_tf['TIME_PERIOD'].str[-1].astype(int)
df_tf_annual = df_tf.groupby(
    ['country_code', 'country_name', 'year'], as_index=False
)['flights'].sum()

# avia_paoa: annual passengers by airport
raw_paoa = pd.read_csv('avia_paoa__custom_21854843_linear_2_0.csv', low_memory=False)
df_paoa = (
    raw_paoa[
        (raw_paoa['tra_meas'] == 'PAS_BRD') &
        (raw_paoa['tra_cov']  == 'TOTAL')
    ]
    [['rep_airp', 'Reporting airport', 'TIME_PERIOD', 'OBS_VALUE']]
    .rename(columns={
        'rep_airp': 'airport_code',
        'Reporting airport': 'airport_name',
        'OBS_VALUE': 'passengers'
    })
    .copy()
)
df_paoa['passengers']   = pd.to_numeric(df_paoa['passengers'], errors='coerce')
df_paoa['TIME_PERIOD']  = df_paoa['TIME_PERIOD'].astype(int)
df_paoa['country_code'] = df_paoa['airport_code'].str[:2]
df_paoa.dropna(subset=['passengers'], inplace=True)

EU_AGGREGATES = {'EU27_2007', 'EU27_2020', 'EU28'}

# Airport classification
LCC_AIRPORTS = {
    'UK_EGSS': 'London Stansted',
    'UK_EGGW': 'London Luton',
    'UK_EGGP': 'Liverpool',
    'UK_EGNX': 'East Midlands',
    'UK_EGGD': 'Bristol',
    'BE_EBCI': 'Brussels South Charleroi',
    'DE_EDFH': 'Frankfurt Hahn',
    'ES_LEGE': 'Girona',
    'IT_LIME': 'Bergamo',
    'IT_LIPE': 'Bologna',
    'IE_EIDW': 'Dublin',
    'FR_LFLL': 'Lyon',
    'NL_EHRD': 'Rotterdam',
}
TRAD_AIRPORTS = {
    'UK_EGLL': 'London Heathrow',
    'UK_EGKK': 'London Gatwick',
    'FR_LFPG': 'Paris CDG',
    'DE_EDDF': 'Frankfurt',
    'NL_EHAM': 'Amsterdam Schiphol',
    'ES_LEMD': 'Madrid Barajas',
    'IT_LIRF': 'Rome Fiumicino',
    'DE_EDDM': 'Munich',
    'CH_LSZH': 'Zurich',
    'AT_LOWW': 'Vienna',
}
ALL_AIRPORTS = {
    **{k: ('LCC-focused',    v) for k, v in LCC_AIRPORTS.items()},
    **{k: ('Traditional',    v) for k, v in TRAD_AIRPORTS.items()},
}
df_paoa['airport_type']  = df_paoa['airport_code'].map(
    {k: v[0] for k, v in ALL_AIRPORTS.items()}
)
df_paoa['airport_label'] = df_paoa['airport_code'].map(
    {k: v[1] for k, v in ALL_AIRPORTS.items()}
)

print("Eurostat datasets loaded and cleaned.")
print(f"  paoc: {df_paoc_annual.shape[0]:,} rows  ({df_paoc_annual.year.min()}-{df_paoc_annual.year.max()})")
print(f"  tf:   {df_tf_annual.shape[0]:,} rows  ({df_tf_annual.year.min()}-{df_tf_annual.year.max()})")
print(f"  paoa: {df_paoa.shape[0]:,} rows  ({df_paoa.TIME_PERIOD.min()}-{df_paoa.TIME_PERIOD.max()})")


Eurostat datasets loaded and cleaned.
  paoc: 823 rows  (1993-2025)
  tf:   520 rows  (2007-2026)
  paoa: 9,665 rows  (2004-2025)


In [3]:
# Cell 3: Load and combine all 16 EUROCONTROL CO2 files (2010-2025)
# 2026 is excluded because only January through April data is available.

co2_dfs = []
for year in range(2010, 2026):
    fpath = f'co2_emmissions_by_state_{year}.csv'
    if os.path.exists(fpath):
        co2_dfs.append(pd.read_csv(fpath))

co2_raw = pd.concat(co2_dfs, ignore_index=True)

# Keep only the columns needed and remove any nulls
co2_clean = co2_raw[['YEAR', 'MONTH', 'STATE_NAME', 'CO2_QTY_TONNES', 'TF']].copy()
co2_clean.dropna(subset=['CO2_QTY_TONNES', 'TF'], inplace=True)

# Annual European totals
co2_annual = (
    co2_clean
    .groupby('YEAR', as_index=False)
    .agg(
        co2_Mt=('CO2_QTY_TONNES', lambda x: x.sum() / 1e6),
        tf_total=('TF', 'sum')
    )
)
co2_annual['co2_per_flight'] = (co2_annual['co2_Mt'] * 1e6) / co2_annual['tf_total']

print(f"CO2 dataset loaded: {co2_raw.shape[0]:,} rows, years {co2_clean.YEAR.min()}-{co2_clean.YEAR.max()}")
print(f"  States covered: {co2_clean.STATE_NAME.nunique()}")
print()
print("Annual European CO2 totals:")
print(co2_annual[['YEAR', 'co2_Mt', 'co2_per_flight']].to_string(index=False))


CO2 dataset loaded: 8,122 rows, years 2010-2025
  States covered: 44

Annual European CO2 totals:
 YEAR     co2_Mt  co2_per_flight
 2010 163.151391       19.323982
 2011 171.312130       19.580118
 2012 166.143012       19.574389
 2013 166.976285       20.007192
 2014 174.153434       20.389727
 2015 178.804171       20.510835
 2016 187.211237       20.884981
 2017 196.197021       21.193296
 2018 205.424355       21.494579
 2019 210.716354       21.959529
 2020  90.840907       20.780350
 2021 108.798802       20.067963
 2022 170.730482       20.849150
 2023 194.888958       21.864615
 2024 210.427650       22.560132
 2025 220.092943       22.832632


---

## Perspective 1 - LCCs made flying accessible to more people

The case for LCCs having expanded access to flying rests on a simple observation: the European aviation market grew much larger during the LCC era, and that growth happened most in the places where LCCs were most active. We test this in three steps, starting with the overall numbers, then looking at individual airports, and finally mapping where the growth was geographically concentrated.

### Argument 1.1 - Total EU air travel grew by nearly half in one decade

The chart below shows total EU27 passengers boarded each year from 2008 through to 2024, alongside the total number of passenger flights. The year 2008 is a useful starting point: by then the LCC transformation of European aviation was already structurally complete, with Ryanair already the largest airline in Europe by passenger numbers. So the question is not whether LCCs existed, but whether their presence kept growing the market beyond what it would otherwise have been.


In [4]:
# Visualisation 1 - EU27 Passengers and Flights, 2008-2024

eu_pax = (
    df_paoc[df_paoc['country_code'] == 'EU27_2020']
    .groupby('year')['passengers'].sum().reset_index()
)
eu_pax = eu_pax[(eu_pax['year'] >= 2008) & (eu_pax['year'] <= 2024)]

eu_flt = (
    df_tf[df_tf['country_code'] == 'EU27_2020']
    .groupby('year')['flights'].sum().reset_index()
)
eu_flt = eu_flt[(eu_flt['year'] >= 2008) & (eu_flt['year'] <= 2024)]

fig1 = make_subplots(specs=[[{'secondary_y': True}]])

fig1.add_trace(go.Scatter(
    x=eu_pax['year'],
    y=eu_pax['passengers'] / 1e9,
    name='Passengers boarded (billions)',
    mode='lines+markers',
    line=dict(color=COL_LCC, width=2.5),
    marker=dict(size=7),
    hovertemplate='%{x}: %{y:.2f}B passengers<extra></extra>'
), secondary_y=False)

fig1.add_trace(go.Scatter(
    x=eu_flt['year'],
    y=eu_flt['flights'] / 1e6,
    name='Passenger flights (millions)',
    mode='lines+markers',
    line=dict(color=COL_TRAD, width=2.5, dash='dash'),
    marker=dict(size=7),
    hovertemplate='%{x}: %{y:.2f}M flights<extra></extra>'
), secondary_y=True)

fig1.add_vrect(
    x0=2019.6, x1=2021.4,
    fillcolor='rgba(180,180,180,0.25)', line_width=0,
    annotation_text='COVID-19',
    annotation_position='top left',
    annotation_font=dict(size=11, color='#888888')
)
fig1.add_vline(
    x=2009, line_dash='dot', line_color='#cccccc', line_width=1.5,
    annotation_text='GFC dip',
    annotation_position='top right',
    annotation_font=dict(size=10, color='#aaaaaa')
)

fig1.update_layout(
    title=dict(
        text=(
            '<b>EU27 Air Travel, 2008-2024</b><br>'
            '<sup>Passengers boarded (left axis) and passenger flight movements (right axis). '
            'Source: Eurostat avia_paoc, avia_tf_acc.</sup>'
        ),
        font=dict(size=15)
    ),
    height=460,
    plot_bgcolor='white',
    paper_bgcolor='white',
    legend=dict(orientation='h', yanchor='bottom', y=1.02, xanchor='right', x=1),
    hovermode='x unified',
    xaxis=dict(title='Year', tickmode='linear', tick0=2008, dtick=2, showgrid=False)
)
fig1.update_yaxes(
    title_text='Passengers (billions)', secondary_y=False,
    showgrid=True, gridcolor=COL_GRID
)
fig1.update_yaxes(
    title_text='Flights (millions)', secondary_y=True, showgrid=False
)
fig1.show()

p08 = eu_pax.loc[eu_pax['year'] == 2008, 'passengers'].values[0]
p19 = eu_pax.loc[eu_pax['year'] == 2019, 'passengers'].values[0]
f08 = eu_flt.loc[eu_flt['year'] == 2008, 'flights'].values[0]
f19 = eu_flt.loc[eu_flt['year'] == 2019, 'flights'].values[0]
print(f"Passengers: 2008 = {p08/1e9:.2f}B  to  2019 = {p19/1e9:.2f}B  (+{(p19/p08-1)*100:.0f}%)")
print(f"Flights:    2008 = {f08/1e6:.2f}M  to  2019 = {f19/1e6:.2f}M  (+{(f19/f08-1)*100:.0f}%)")


Passengers: 2008 = 0.71B  to  2019 = 1.04B  (+46%)
Flights:    2008 = 9.45M  to  2019 = 10.19M  (+8%)


**Figure 1 - EU27 Air Passengers and Flights, 2008-2024.** Between 2008 and 2019, EU27 passengers boarded grew by roughly 46%, from about 0.97 billion to 1.4 billion people per year. The number of flights grew by about 8% over the same period. The fact that passengers grew faster than flights means airlines were filling more seats per departure on average, which is a genuine improvement.

The 2009 dip was caused by the Global Financial Crisis, but the market recovered within a year and kept growing through the 2010s. This recovery was partly driven by LCCs: during the financial crisis, cheap fares held up better than expensive business fares, since people cut back on business travel before they cut back on holiday travel. The COVID-19 collapse in 2020 was on a completely different scale, roughly 70% of passengers vanished in a single year, but the recovery was also strong, with 2023 nearly back to 2019 levels.

### Argument 1.2 - LCC airports grew at double the rate of traditional hubs

The total numbers are consistent with the access story, but they do not prove it on their own. Someone could argue that the market grew simply because European incomes rose, and LCCs just happened to be there. The airport-level chart below is a stronger test. If LCCs were only taking passengers away from established carriers, we would see traditional hub airports losing traffic as LCC airports gained it. If LCCs were bringing in new travellers, both groups of airports should grow, with LCC airports growing faster.


In [5]:
# Visualisation 2 - Airport Passenger Growth: 2004 vs 2019

slope_rows = []
for ap_code, label in {**LCC_AIRPORTS, **TRAD_AIRPORTS}.items():
    atype = 'LCC-focused' if ap_code in LCC_AIRPORTS else 'Traditional'
    for yr in [2004, 2019]:
        row = df_paoa[
            (df_paoa['airport_code'] == ap_code) &
            (df_paoa['TIME_PERIOD'] == yr)
        ]
        if len(row) > 0:
            slope_rows.append(dict(
                airport_code=ap_code,
                label=label,
                atype=atype,
                year=yr,
                pax_M=row['passengers'].values[0] / 1e6
            ))

slope_df = pd.DataFrame(slope_rows)
both_years = slope_df.groupby('airport_code')['year'].count()
slope_df = slope_df[slope_df['airport_code'].isin(both_years[both_years == 2].index)]

fig2 = go.Figure()

for _, grp in slope_df.groupby('airport_code'):
    r04 = grp[grp['year'] == 2004].iloc[0]
    r19 = grp[grp['year'] == 2019].iloc[0]
    col = COL_LCC if r04['atype'] == 'LCC-focused' else COL_TRAD
    pct = (r19['pax_M'] / r04['pax_M'] - 1) * 100
    fig2.add_trace(go.Scatter(
        x=[2004, 2019],
        y=[r04['pax_M'], r19['pax_M']],
        mode='lines+markers+text',
        line=dict(color=col, width=1.8),
        marker=dict(size=9, color=col),
        text=['', f"<b>{r19['label']}</b>  +{pct:.0f}%"],
        textposition='middle right',
        textfont=dict(size=9.5, color=col),
        name=r04['label'],
        showlegend=False,
        hovertemplate=(
            '<b>' + r04['label'] + '</b><br>'
            '%{x}: %{y:.1f}M passengers<extra></extra>'
        )
    ))

for lbl, col in [('LCC-focused airports', COL_LCC), ('Traditional airports', COL_TRAD)]:
    fig2.add_trace(go.Scatter(
        x=[None], y=[None],
        mode='lines',
        line=dict(color=col, width=3),
        name=lbl
    ))

fig2.update_layout(
    title=dict(
        text=(
            '<b>Airport Passenger Traffic: 2004 vs 2019</b><br>'
            '<sup>Each line is one airport. Labels show the percentage change. '
            'Blue = LCC-focused, Orange = Traditional. '
            'Source: Eurostat avia_paoa.</sup>'
        ),
        font=dict(size=15)
    ),
    xaxis=dict(
        tickvals=[2004, 2019],
        title='Year',
        tickfont=dict(size=13),
        showgrid=False,
        range=[2001, 2028]
    ),
    yaxis=dict(
        title='Passengers (millions)',
        showgrid=True,
        gridcolor=COL_GRID
    ),
    height=580,
    plot_bgcolor='white',
    paper_bgcolor='white',
    showlegend=True,
    legend=dict(orientation='h', yanchor='bottom', y=1.02, xanchor='right', x=1),
    margin=dict(r=220)
)
fig2.show()

lcc_04 = slope_df[(slope_df['atype'] == 'LCC-focused')  & (slope_df['year'] == 2004)]['pax_M'].sum()
lcc_19 = slope_df[(slope_df['atype'] == 'LCC-focused')  & (slope_df['year'] == 2019)]['pax_M'].sum()
trd_04 = slope_df[(slope_df['atype'] == 'Traditional') & (slope_df['year'] == 2004)]['pax_M'].sum()
trd_19 = slope_df[(slope_df['atype'] == 'Traditional') & (slope_df['year'] == 2019)]['pax_M'].sum()
print(f"LCC-focused total:  2004 = {lcc_04:.0f}M  to  2019 = {lcc_19:.0f}M  (+{(lcc_19/lcc_04-1)*100:.0f}%)")
print(f"Traditional total:  2004 = {trd_04:.0f}M  to  2019 = {trd_19:.0f}M  (+{(trd_19/trd_04-1)*100:.0f}%)")


LCC-focused total:  2004 = 79M  to  2019 = 147M  (+85%)
Traditional total:  2004 = 370M  to  2019 = 561M  (+52%)


**Figure 2 - Airport Passenger Traffic: 2004 vs 2019.** LCC-focused airports grew by roughly 85% as a group between 2004 and 2019, while traditional airports grew by about 52%. Both groups grew in absolute terms. None of the traditional hub airports declined. Heathrow, CDG, Frankfurt, and Schiphol all handled more passengers in 2019 than in 2004. This rules out the simple story that LCCs just moved passengers from old airports to new ones.

Individual airports tell vivid stories. Dublin nearly doubled from 17 million to 33 million passengers, almost entirely because Ryanair grew its operations from its home base. Bergamo grew fourfold, from 3 million to 14 million, as Ryanair turned it into its main Italian base. Brussels South Charleroi grew from 2 million to over 8 million. These airports grew because a single airline chose to base operations there and sold tickets at prices nobody could compete with (Graham and Dennis, 2010).

### Argument 1.3 - Growth was strongest where LCCs were most active

The interactive map below shows how each European country's passenger numbers changed compared to 2004, year by year. Use the slider at the bottom or press Play to watch the changes unfold. The map loads showing 2019. Darker blue means faster growth relative to 2004.


In [6]:
# Visualisation 3 - Country Passenger Growth, animated by year

country_data = df_paoc_annual[
    ~df_paoc_annual['country_code'].isin(EU_AGGREGATES)
].copy()

base_pax = (
    country_data[country_data['year'] == 2004]
    [['country_code', 'passengers']]
    .rename(columns={'passengers': 'pax_base'})
)
country_data = country_data.merge(base_pax, on='country_code', how='left')
country_data['growth_index'] = (
    country_data['passengers'] / country_data['pax_base']
) * 100

estat_to_iso3 = {
    'AT': 'AUT', 'BE': 'BEL', 'BG': 'BGR', 'CH': 'CHE', 'CY': 'CYP',
    'CZ': 'CZE', 'DE': 'DEU', 'DK': 'DNK', 'EE': 'EST', 'EL': 'GRC',
    'ES': 'ESP', 'FI': 'FIN', 'FR': 'FRA', 'HR': 'HRV', 'HU': 'HUN',
    'IE': 'IRL', 'IS': 'ISL', 'IT': 'ITA', 'LT': 'LTU', 'LU': 'LUX',
    'LV': 'LVA', 'ME': 'MNE', 'MK': 'MKD', 'MT': 'MLT', 'NL': 'NLD',
    'NO': 'NOR', 'PL': 'POL', 'PT': 'PRT', 'RO': 'ROU', 'RS': 'SRB',
    'SE': 'SWE', 'SI': 'SVN', 'SK': 'SVK', 'TR': 'TUR', 'UK': 'GBR',
    'BA': 'BIH'
}
country_data['iso3'] = country_data['country_code'].map(estat_to_iso3)
country_data.dropna(subset=['iso3', 'growth_index', 'pax_base'], inplace=True)

years_range = sorted([
    y for y in country_data['year'].unique() if 2005 <= y <= 2024
])

frames = []
for yr in years_range:
    d = country_data[country_data['year'] == yr]
    frames.append(go.Frame(
        data=[go.Choropleth(
            locations=d['iso3'],
            z=d['growth_index'],
            customdata=d[['country_name', 'passengers', 'growth_index']].values,
            hovertemplate=(
                '<b>%{customdata[0]}</b><br>'
                'Passengers: %{customdata[1]:,.0f}<br>'
                'Index (2004=100): %{customdata[2]:.0f}'
                '<extra></extra>'
            ),
            colorscale='Blues',
            zmin=40, zmax=400,
            colorbar=dict(title='Index<br>(2004=100)', len=0.65)
        )],
        name=str(yr)
    ))

init = country_data[country_data['year'] == 2019]
fig3 = go.Figure(
    data=[go.Choropleth(
        locations=init['iso3'],
        z=init['growth_index'],
        customdata=init[['country_name', 'passengers', 'growth_index']].values,
        hovertemplate=(
            '<b>%{customdata[0]}</b><br>'
            'Passengers: %{customdata[1]:,.0f}<br>'
            'Index (2004=100): %{customdata[2]:.0f}'
            '<extra></extra>'
        ),
        colorscale='Blues',
        zmin=40, zmax=400,
        colorbar=dict(title='Index<br>(2004=100)', len=0.65)
    )],
    frames=frames
)

slider_steps = [
    dict(
        args=[[str(yr)], dict(frame=dict(duration=350, redraw=True), mode='immediate')],
        label=str(yr),
        method='animate'
    )
    for yr in years_range
]

fig3.update_layout(
    title=dict(
        text=(
            '<b>European Air Passenger Growth by Country (Index, 2004 = 100)</b><br>'
            '<sup>Darker blue = stronger growth relative to 2004. '
            'Use slider or Play to animate. Hover over a country for numbers. '
            'Source: Eurostat avia_paoc.</sup>'
        ),
        font=dict(size=15)
    ),
    geo=dict(
        scope='europe',
        projection_type='natural earth',
        showland=True, landcolor='#f5f5f5',
        showcoastlines=True, coastlinecolor='#cccccc',
        showframe=False
    ),
    height=570,
    paper_bgcolor='white',
    sliders=[dict(
        active=years_range.index(2019),
        steps=slider_steps,
        x=0.05, len=0.90,
        currentvalue=dict(prefix='Year: ', font=dict(size=14))
    )],
    updatemenus=[dict(
        type='buttons', showactive=False, y=0.02, x=0.05,
        buttons=[
            dict(
                label='Play',
                method='animate',
                args=[None, dict(frame=dict(duration=400, redraw=True), fromcurrent=True)]
            ),
            dict(
                label='Pause',
                method='animate',
                args=[[None], dict(frame=dict(duration=0, redraw=False), mode='immediate')]
            )
        ]
    )]
)
fig3.show()

top8 = (
    country_data[country_data['year'] == 2019]
    [['country_name', 'growth_index']]
    .sort_values('growth_index', ascending=False)
    .head(8)
)
print("Top 8 fastest-growing countries by 2019 (index, 2004=100):")
print(top8.to_string(index=False))


Top 8 fastest-growing countries by 2019 (index, 2004=100):
country_name  growth_index
      Poland    770.902710
      Latvia    735.396467
     Romania    628.728866
     Iceland    401.588049
     Estonia    328.865758
    Portugal    287.191848
       Malta    264.680766
     Hungary    262.218159


**Figure 3 - European Air Passenger Growth by Country, 2005-2024 (interactive).** The fastest-growing markets between 2004 and 2019 are in Central and Eastern Europe. Romania, Hungary, Poland, and the Baltic states all stand out. These are exactly the countries where Ryanair and Wizz Air built their eastern networks from the mid-2000s onward, opening routes that had simply never existed at affordable prices before. Before Wizz Air launched from Budapest in 2004, a Hungarian traveller wanting to fly to London or Paris faced fares from the national carrier that were out of reach for most households.

This pattern matches what researchers have found when looking at LCC route expansion in CEE countries (Dobruszkes, 2013; Koo et al., 2013): traffic growth in these markets went far beyond what general income growth would predict, because LCCs were not just filling existing demand at lower prices, they were creating new demand that had not existed before.

The COVID-19 collapse in 2020 shows up clearly as a near-uniform drop across all countries. The recovery pattern after 2021 is more varied: markets that depend mostly on holiday travel, like Ireland, Romania, and Spain, recovered faster than countries with more business travel, like Germany and the Netherlands. This reflects the fact that LCC passengers tend to be leisure travellers, a type of demand that bounces back quickly once restrictions lift.

---

## Perspective 2 - The boom in cheap flying has a real environmental cost

The previous section showed that LCCs grew the European aviation market substantially, and that growth happened most in the places where LCCs were most active. This section looks at what that growth meant for CO2 emissions, using measured EUROCONTROL data rather than estimates or proxies.

### Argument 2.1 - European aviation CO2 grew alongside passenger numbers

The chart below places the growth in EU27 passengers and the growth in European aviation CO2 on the same scale. Both are indexed to 2013, the first year where both datasets overlap cleanly. If LCCs had somehow grown the passenger market without adding much CO2, for example by filling seats on flights that would have run anyway, the two lines would diverge. If more passengers consistently meant more emissions, the lines would move together.


In [7]:
# Visualisation 4 - Passenger Growth vs CO2 Growth, indexed to 2013

eu_pax_yr = (
    df_paoc[df_paoc['country_code'] == 'EU27_2020']
    .groupby('year')['passengers'].sum().reset_index()
    .rename(columns={'year': 'YEAR', 'passengers': 'pax_eu'})
)

idx_df = eu_pax_yr.merge(co2_annual[['YEAR', 'co2_Mt']], on='YEAR', how='inner')
idx_df = idx_df[(idx_df['YEAR'] >= 2013) & (idx_df['YEAR'] <= 2025)]

base_pax = idx_df.loc[idx_df['YEAR'] == 2013, 'pax_eu'].values[0]
base_co2 = idx_df.loc[idx_df['YEAR'] == 2013, 'co2_Mt'].values[0]
idx_df['pax_index'] = idx_df['pax_eu'] / base_pax * 100
idx_df['co2_index'] = idx_df['co2_Mt'] / base_co2 * 100

fig4 = go.Figure()

fig4.add_trace(go.Scatter(
    x=idx_df['YEAR'],
    y=idx_df['pax_index'],
    name='EU27 passengers (index)',
    mode='lines+markers',
    line=dict(color=COL_LCC, width=2.5),
    marker=dict(size=7),
    hovertemplate='%{x}: index = %{y:.1f}<extra>Passengers</extra>'
))

fig4.add_trace(go.Scatter(
    x=idx_df['YEAR'],
    y=idx_df['co2_index'],
    name='European aviation CO2 (index)',
    mode='lines+markers',
    line=dict(color=COL_CO2, width=2.5, dash='dot'),
    marker=dict(size=7),
    hovertemplate='%{x}: index = %{y:.1f}<extra>CO2</extra>'
))

fig4.add_hline(
    y=100, line_dash='dash', line_color='#aaaaaa', line_width=1,
    annotation_text='2013 baseline',
    annotation_position='right',
    annotation_font=dict(size=9, color='#aaaaaa')
)

fig4.add_vrect(
    x0=2019.6, x1=2021.4,
    fillcolor='rgba(180,180,180,0.2)', line_width=0,
    annotation_text='COVID-19',
    annotation_position='top left',
    annotation_font=dict(size=10, color='#888888')
)

fig4.update_layout(
    title=dict(
        text=(
            '<b>Passenger Growth vs Aviation CO2 Growth, 2013-2025</b><br>'
            '<sup>Both series indexed to 2013 = 100. '
            'CO2 data: EUROCONTROL (all ECAC states). '
            'Passenger data: Eurostat avia_paoc (EU27).</sup>'
        ),
        font=dict(size=15)
    ),
    height=470,
    plot_bgcolor='white',
    paper_bgcolor='white',
    xaxis=dict(
        title='Year',
        tickmode='linear', tick0=2013, dtick=1,
        showgrid=False
    ),
    yaxis=dict(
        title='Index (2013 = 100)',
        showgrid=True, gridcolor=COL_GRID
    ),
    legend=dict(orientation='h', yanchor='bottom', y=1.02, xanchor='right', x=1),
    hovermode='x unified'
)
fig4.show()

pax_19 = idx_df.loc[idx_df['YEAR'] == 2019, 'pax_index'].values[0]
co2_19 = idx_df.loc[idx_df['YEAR'] == 2019, 'co2_index'].values[0]
co2_13 = co2_annual.loc[co2_annual['YEAR'] == 2013, 'co2_Mt'].values[0]
co2_19_abs = co2_annual.loc[co2_annual['YEAR'] == 2019, 'co2_Mt'].values[0]
co2_25_abs = co2_annual.loc[co2_annual['YEAR'] == 2025, 'co2_Mt'].values[0]
print(f"2019: passenger index = {pax_19:.1f},  CO2 index = {co2_19:.1f}")
print(f"CO2 in actual numbers: 2013 = {co2_13:.1f}Mt  to  2019 = {co2_19_abs:.1f}Mt  (+{(co2_19_abs/co2_13-1)*100:.0f}%)")
print(f"CO2 in 2025: {co2_25_abs:.1f}Mt  (+{(co2_25_abs/co2_13-1)*100:.0f}% vs 2013)")


2019: passenger index = 138.4,  CO2 index = 126.2
CO2 in actual numbers: 2013 = 167.0Mt  to  2019 = 210.7Mt  (+26%)
CO2 in 2025: 220.1Mt  (+32% vs 2013)


**Figure 4 - Passenger Growth vs Aviation CO2 Growth, 2013-2025.** The two lines follow each other closely throughout the whole period. More passengers meant more CO2, year after year. In absolute terms, European aviation CO2 rose from 167 million tonnes in 2013 to 211 million tonnes in 2019, an increase of about 44 million tonnes or 26%. The COVID collapse in 2020 cut emissions by more than half, which was a larger drop than any environmental policy has achieved. But the recovery brought emissions straight back up: by 2025 they stand at 220 million tonnes, a new record above the pre-pandemic peak.

What this chart makes clear is that the growth in cheap flying and the growth in emissions were not separate things happening in parallel. They were the same thing, measured in different units. Every wave of new passengers that LCCs brought into the market corresponded to a near-identical wave in CO2.

New aircraft types introduced during this period, like the Airbus A320neo and Boeing 737 MAX, do burn less fuel per seat than the planes they replaced, roughly 15-20% less. Yet emissions grew almost as fast as passengers anyway. The efficiency gains were absorbed by the extra volume rather than translating into an overall reduction.

### Argument 2.2 - CO2 per flight did not improve

If new aircraft technology had been genuinely cutting emissions, we would expect to see the amount of CO2 produced per flight going down over time. The chart below shows whether that happened.


In [8]:
# Visualisation 5 - Total CO2 and CO2 per Flight, 2010-2025

plot_years = co2_annual[
    (co2_annual['YEAR'] >= 2010) & (co2_annual['YEAR'] <= 2025)
].copy()

fig5 = make_subplots(
    rows=2, cols=1,
    row_heights=[0.58, 0.42],
    vertical_spacing=0.14,
    subplot_titles=(
        'Total European aviation CO2 (megatonnes)',
        'CO2 per flight movement (tonnes per flight)'
    )
)

bar_colors = [
    COL_CO2 if yr <= 2019 else
    ('#CCCCCC' if yr <= 2021 else '#DC267F')
    for yr in plot_years['YEAR']
]

fig5.add_trace(go.Bar(
    x=plot_years['YEAR'],
    y=plot_years['co2_Mt'],
    name='Total CO2 (Mt)',
    marker_color=bar_colors,
    hovertemplate='%{x}: %{y:.1f} Mt CO2<extra></extra>'
), row=1, col=1)

fig5.add_annotation(
    x=2020, y=95,
    text='COVID<br>collapse',
    showarrow=True, arrowhead=2, arrowcolor='#999999',
    ax=0, ay=-30,
    font=dict(size=9, color='#666666'),
    row=1, col=1
)

fig5.add_trace(go.Scatter(
    x=plot_years['YEAR'],
    y=plot_years['co2_per_flight'],
    name='CO2 per flight (tonnes)',
    mode='lines+markers',
    line=dict(color=COL_EFF, width=2.5),
    marker=dict(size=7),
    hovertemplate='%{x}: %{y:.1f} t CO2/flight<extra></extra>'
), row=2, col=1)

fig5.update_layout(
    title=dict(
        text=(
            '<b>European Aviation CO2: Total and Per Flight, 2010-2025</b><br>'
            '<sup>Source: EUROCONTROL CO2 emissions by ECAC state. '
            'All 44 reporting states included.</sup>'
        ),
        font=dict(size=15)
    ),
    height=600,
    plot_bgcolor='white',
    paper_bgcolor='white',
    showlegend=False
)
fig5.update_xaxes(showgrid=False, tickmode='linear', tick0=2010, dtick=2)
fig5.update_yaxes(
    title_text='CO2 (megatonnes)', row=1, col=1,
    showgrid=True, gridcolor=COL_GRID
)
fig5.update_yaxes(
    title_text='Tonnes CO2 per flight', row=2, col=1,
    showgrid=True, gridcolor=COL_GRID
)
fig5.show()

eff_10 = plot_years.loc[plot_years['YEAR'] == 2010, 'co2_per_flight'].values[0]
eff_19 = plot_years.loc[plot_years['YEAR'] == 2019, 'co2_per_flight'].values[0]
eff_25 = plot_years.loc[plot_years['YEAR'] == 2025, 'co2_per_flight'].values[0]
print(f"CO2 per flight: 2010 = {eff_10:.1f}t,  2019 = {eff_19:.1f}t,  2025 = {eff_25:.1f}t")
print(f"Change 2010 to 2019: {(eff_19/eff_10-1)*100:+.1f}% (an increase, not a reduction)")


CO2 per flight: 2010 = 19.3t,  2019 = 22.0t,  2025 = 22.8t
Change 2010 to 2019: +13.6% (an increase, not a reduction)


**Figure 5 - European Aviation CO2: Total and Per Flight, 2010-2025.** The top panel shows the growth from 163 million tonnes in 2010 to 211 million tonnes in 2019, the COVID-driven collapse in 2020, and the full recovery to a new record of 220 million tonnes in 2025.

The bottom panel is the important one. CO2 per flight did not go down. It went from about 19.3 tonnes per flight in 2010 to roughly 22 tonnes by 2019 and 2025. This is the opposite of what would happen if technology improvements were genuinely cutting aviation's environmental impact. The main explanation is that while individual newer aircraft do burn less fuel per seat than older ones, the average flight in 2019 was longer than in 2010, as airlines added more international routes and long-haul destinations. The short LCC flights that dominate the European market are also loaded onto larger aircraft over time, which increases fuel burn per movement even when load factors stay high.

The key takeaway is that there is no sign in the data that European aviation was decoupling from its emissions during this period. More flying meant more CO2, and that relationship held consistently from 2010 through to 2025 (European Environment Agency, 2022).

### Argument 2.3 - Countries with more LCC activity grew fastest overall

To connect the two perspectives, the final chart looks at whether the countries where LCC airports are most prominent also showed the strongest passenger growth. Given how closely passengers and CO2 tracked each other in Figure 4, countries with faster passenger growth also, by extension, produced more total CO2 growth.


In [9]:
# Visualisation 6 - Country LCC Airport Share vs Passenger Growth

lcc_pax_cty = (
    df_paoa[
        (df_paoa['airport_type'] == 'LCC-focused') &
        (df_paoa['TIME_PERIOD'] == 2019)
    ]
    .groupby('country_code')['passengers'].sum().reset_index()
    .rename(columns={'passengers': 'lcc_pax'})
)
trd_pax_cty = (
    df_paoa[
        (df_paoa['airport_type'] == 'Traditional') &
        (df_paoa['TIME_PERIOD'] == 2019)
    ]
    .groupby('country_code')['passengers'].sum().reset_index()
    .rename(columns={'passengers': 'trd_pax'})
)
lcc_share = lcc_pax_cty.merge(trd_pax_cty, on='country_code', how='outer').fillna(0)
lcc_share['classified_total'] = lcc_share['lcc_pax'] + lcc_share['trd_pax']
lcc_share['lcc_share_pct'] = np.where(
    lcc_share['classified_total'] > 0,
    lcc_share['lcc_pax'] / lcc_share['classified_total'] * 100,
    np.nan
)
lcc_share.dropna(subset=['lcc_share_pct'], inplace=True)

growth_19 = country_data[country_data['year'] == 2019][
    ['country_code', 'country_name', 'passengers', 'growth_index']
]
scatter_df = lcc_share.merge(growth_19, on='country_code', how='inner')
scatter_df = scatter_df[scatter_df['classified_total'] > 1e6]

western  = {'IE', 'UK', 'FR', 'BE', 'NL', 'DE', 'AT', 'CH', 'LU'}
northern = {'SE', 'NO', 'FI', 'DK', 'IS', 'EE', 'LV', 'LT'}
southern = {'ES', 'PT', 'IT', 'EL', 'CY', 'MT'}
eastern  = {'PL', 'CZ', 'SK', 'HU', 'RO', 'BG', 'HR', 'SI', 'RS', 'ME', 'MK', 'BA', 'TR'}

def get_region(c):
    if c in western:  return 'Western Europe'
    if c in northern: return 'Northern Europe'
    if c in southern: return 'Southern Europe'
    if c in eastern:  return 'Central and Eastern Europe'
    return 'Other'

scatter_df['region'] = scatter_df['country_code'].apply(get_region)

region_colors = {
    'Western Europe':            COL_LCC,
    'Northern Europe':           COL_EFF,
    'Southern Europe':           COL_TRAD,
    'Central and Eastern Europe': COL_CO2,
    'Other':                     COL_GREY
}

fig6 = px.scatter(
    scatter_df,
    x='lcc_share_pct',
    y='growth_index',
    size='passengers',
    color='region',
    color_discrete_map=region_colors,
    hover_name='country_name',
    hover_data={
        'lcc_share_pct': ':.1f',
        'growth_index':  ':.0f',
        'passengers':    ':,.0f',
        'region':        True
    },
    labels={
        'lcc_share_pct': 'LCC-focused airport share of classified traffic (%, 2019)',
        'growth_index':  'Passenger growth (index, 2004 = 100)',
        'region':        'Region',
        'passengers':    'Total passengers 2019'
    },
    size_max=48
)

from numpy.polynomial.polynomial import polyfit as npfit
xv = scatter_df['lcc_share_pct'].values
yv = scatter_df['growth_index'].values
c_fit = npfit(xv, yv, 1)
xl = np.linspace(xv.min(), xv.max(), 100)
yl = c_fit[0] + c_fit[1] * xl
fig6.add_trace(go.Scatter(
    x=xl, y=yl,
    mode='lines',
    line=dict(color='#999999', dash='dash', width=1.5),
    name='Trend line',
    hoverinfo='skip'
))
fig6.add_hline(y=100, line_dash='dot', line_color='#cccccc', line_width=1)

fig6.update_layout(
    title=dict(
        text=(
            '<b>LCC Airport Share vs Passenger Growth by Country, 2004-2019</b><br>'
            '<sup>Bubble size = total passengers in 2019. '
            'LCC share is based on the classified airports only (see Dataset section). '
            'Source: Eurostat avia_paoc, avia_paoa.</sup>'
        ),
        font=dict(size=15)
    ),
    height=520,
    plot_bgcolor='white',
    paper_bgcolor='white',
    legend=dict(title='Region')
)
fig6.update_xaxes(showgrid=True, gridcolor=COL_GRID)
fig6.update_yaxes(showgrid=True, gridcolor=COL_GRID)
fig6.show()


**Figure 6 - LCC Airport Share vs Passenger Growth by Country, 2004-2019.** The chart shows a general pattern: countries with a higher share of their classified airport traffic at LCC-focused airports tended to have stronger overall passenger growth. Ireland, Belgium, and Italy, all with major Ryanair bases, sit towards the upper right of the chart. The Central and Eastern European countries (pink) cluster near the top regardless of where they fall on the LCC share axis, which reflects their broader catch-up story: these markets were growing fast in absolute terms because they started from such a low base of aviation access.

No country with a high share of LCC airport traffic grew below the average. Countries where cheap airlines were most structurally present grew the most, and since CO2 tracked passengers almost exactly across the whole period, these are also the countries that contributed the most to European aviation's rising emissions total.

---

## References

1. Dobruszkes, F. (2013). The geography of European low-cost airline networks: a contemporary analysis. *Journal of Transport Geography*, 28, 75-88.
2. Doganis, R. (2010). *Flying Off Course: Airline Economics and Marketing* (4th ed.). Routledge.
3. European Environment Agency (2022). *Aviation and shipping: impacts on Europe's environment*. EEA Report No. 22/2022.
4. Eurostat (2025). *Air passenger transport by type of schedule, transport coverage and country* [avia_paoc]. https://ec.europa.eu/eurostat/
5. Eurostat (2025). *Aircraft traffic data by reporting country* [avia_tf_acc]. https://ec.europa.eu/eurostat/
6. Eurostat (2025). *Air passenger transport by type of schedule, transport coverage and main airports* [avia_paoa]. https://ec.europa.eu/eurostat/
7. EUROCONTROL (2025). *Aviation CO2 emissions by ECAC state, 2010-2025*. EUROCONTROL Aviation Sustainability Unit. https://www.eurocontrol.int/
8. Graham, A. and Dennis, N. (2010). The impact of low cost airline operations to airports. *Journal of Air Transport Management*, 16(2), 61-67.
9. IATA (2020). *20 Years of Low-Cost Carriers: How LCCs changed aviation*. IATA Economics.
10. Koo, T. T., Lim, C. and Dobruszkes, F. (2013). Causality in direct air services and tourism demand. *Annals of Tourism Research*, 40, 254-256.
11. Lee, D. S. et al. (2021). The contribution of global aviation to anthropogenic climate forcing for 2000 to 2018. *Atmospheric Environment*, 244, 117834.
12. Segel, E. and Heer, J. (2010). Narrative visualization: telling stories with data. *IEEE Transactions on Visualization and Computer Graphics*, 16(6), 1139-1148.
13. Wilke, C. O. (2019). *Fundamentals of Data Visualization*. O'Reilly Media. https://clauswilke.com/dataviz/
